File: app/src/main/java/com/plbear/iweight/view/LineChartView.java
package com.plbear.iweight.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Point;
import android.graphics.Typeface;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.plbear.iweight.Data.Data;
import com.plbear.iweight.Data.DataManager;
import com.plbear.iweight.R;
import com.plbear.iweight.Utils.ILog;
import com.plbear.iweight.Utils.OldUtils;

import java.util.ArrayList;

/**
 * Created by koakira on 16/11/5.
 */

public class LineChartView extends View {
    private String TAG = "LineChartView:";
    private final int SPACES_COUNT = 5;
    private final int MSG_DATA_CHANGE = 0;
    private final int ALIGN_PADDING_LEFT = 20;
    private final int ALIGN_PADDING_RIGHT = 20;
    private final int ALIGN_PADDING_TOP = 10;
    private final int ALIGN_PADDING_BOTTOM = 40;

    private Paint mBondLinePaint = null;
    private Paint mLinePaint = null;
    private Paint mDottedLinePaint = null;
    private Paint mPolyLinePaint = null;
    private Paint mTextPaint = null;
    private Paint mPointPaint = null;
    private Paint mTargetPaint = null;
    private Paint mTargetTextPaint = null;
    private DataManager mDataManager = null;

    private int mTop = 0;
    private int mLeft = 0;
    private int mRight = 0;
    private int mBottom = 0;
    private int mWidth = 0;
    private int mHeight = 0;
    private int mLineSpacing = 0;
    private int mCurMoveLength = 0;


    private DataAdapter mDataAdapter = null;
    private Context mContext = null;
    private Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case MSG_DATA_CHANGE:
                    ILog.d(TAG, "Msg_data_change");
                    LineChartView.this.invalidate();
                    break;
                default:
                    break;
            }
        }
    };

    public void setTag(String tag){
        TAG += tag;
        mDataAdapter.setTag(TAG);
    }

    /*public void setDataAdapter(DataAdapter adapter) {
        mDataAdapter = adapter;
        Message msg = mHandler.obtainMessage(MSG_DATA_CHANGE);
        mHandler.sendMessage(msg);
    }*/
    public DataAdapter getDataAdpater() {
        return mDataAdapter;
    }

    public LineChartView(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
        mContext = context;
        init();
    }

    /**
     * 初始化画笔工具
     */
    private void initPaints() {
        mBondLinePaint = new Paint();
        mBondLinePaint.setColor(Color.BLACK);
        mBondLinePaint.setTypeface(Typeface.DEFAULT_BOLD);
        mBondLinePaint.setStyle(Paint.Style.STROKE);
        mBondLinePaint.setStrokeWidth(3);
        mBondLinePaint.setAntiAlias(true);
        mBondLinePaint.setAlpha(130);

        mLinePaint = new Paint();
        mLinePaint.setColor(Color.BLACK);
        mLinePaint.setStyle(Paint.Style.STROKE);
        mLinePaint.setStrokeWidth(3);
        mLinePaint.setAlpha(90);
        mLinePaint.setAntiAlias(true);

        mDottedLinePaint = new Paint();
        mDottedLinePaint.setColor(Color.BLACK);
        mDottedLinePaint.setStyle(Paint.Style.STROKE);
        mDottedLinePaint.setStrokeWidth(3);
        mDottedLinePaint.setAntiAlias(true);
        mDottedLinePaint.setAlpha(150);

        mPolyLinePaint = new Paint();
        mPolyLinePaint.setColor(getResources().getColor(R.color.line_color));
        mPolyLinePaint.setStyle(Paint.Style.STROKE);
        mPolyLinePaint.setStrokeWidth(5);
        mPolyLinePaint.setAntiAlias(true);

        mTextPaint = new Paint();
        mTextPaint.setColor(Color.BLACK);
        mTextPaint.setTextSize(38);
        mTextPaint.setAntiAlias(true);
        mTextPaint.setStyle(Paint.Style.FILL);
        mTextPaint.setAlpha(150);
        mTextPaint.setStrokeWidth(1);

        mPointPaint = new Paint();
        mPointPaint.setColor(getResources().getColor(R.color.point_color));
        mPointPaint.setStrokeWidth(20);

        mTargetPaint = new Paint();
        mTargetPaint.setColor(getResources().getColor(R.color.target_weight));
        mTargetPaint.setStyle(Paint.Style.STROKE);
        mTargetPaint.setStrokeWidth(4);
        mTargetPaint.setAntiAlias(true);

        mTargetTextPaint = new Paint();
        mTargetTextPaint.setColor(getResources().getColor(R.color.target_weight));
        mTargetTextPaint.setTextSize(38);
        mTargetTextPaint.setAntiAlias(true);
        mTargetTextPaint.setStyle(Paint.Style.FILL);
        mTargetTextPaint.setAlpha(150);
        mTargetTextPaint.setStrokeWidth(1);

    }

    private void init() {
        mDataAdapter = new DataAdapter(mContext);
        ILog.e(TAG,"init");
        mDataAdapter.setTag(TAG);
        mDataAdapter.registerDataListener(new DataAdapter.DataChangeListener() {
            @Override
            public void onChange() {
                ILog.d(TAG, "LineChartView data changed");
                Message msg = mHandler.obtainMessage(MSG_DATA_CHANGE);
                mHandler.sendMessage(msg);
            }
        });
        initPaints();
        mDataManager = DataManager.getInstance(mContext);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        if (changed) {
            mLeft = left;
            mTop = top;
            mRight = right;
            mBottom = bottom;
            mWidth = mRight - mLeft - ALIGN_PADDING_LEFT - ALIGN_PADDING_RIGHT;
            mHeight = mBottom - mTop - ALIGN_PADDING_TOP - ALIGN_PADDING_BOTTOM;
            ILog.d(TAG, "mLeft:" + mLeft + " mRight:" + mRight + " mTop:" + mTop + " mBottom:" + mBottom + " mWeight:" + mWidth + " mHeight:" + mHeight);
            mLineSpacing = (mBottom - mTop - ALIGN_PADDING_BOTTOM - ALIGN_PADDING_TOP) / SPACES_COUNT;
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        ILog.d(TAG, "onDraw");
        canvas.drawLine(ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP, mRight - ALIGN_PADDING_RIGHT,
                ALIGN_PADDING_TOP, mBondLinePaint);
        for (int i = 1; i < SPACES_COUNT; i++) {
            canvas.drawLine(ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP + i * mLineSpacing, mRight -
                    ALIGN_PADDING_RIGHT, ALIGN_PADDING_TOP + i * mLineSpacing, mLinePaint);
        }
        canvas.drawLine(ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP + SPACES_COUNT * mLineSpacing, mRight -
                ALIGN_PADDING_RIGHT, ALIGN_PADDING_TOP + SPACES_COUNT * mLineSpacing, mBondLinePaint);

        drawDottedLine(canvas);
        drawPolyLine(canvas);
        drawWeightText(canvas);
        drawTimeText(canvas);
        int temp = mDataAdapter.getShowPointCount();
        ILog.e(TAG,"onDraw:"+temp);
        if(mDataAdapter.getShowPointCount() < 50){
            drawPoint(canvas);
        }
        drawTargetLine(canvas);

        super.onDraw(canvas);
    }


    /**
     * 画目标体重的线和“目标值”这几个数值
     *
     * @param canvas
     */
    private void drawTargetLine(Canvas canvas) {
        float targetWeight = mDataAdapter.getTargetWeight();
        if (targetWeight == -1) {
            ILog.e(TAG, "target weight is not set return");
            return;
        }
        int y = toYPoint(targetWeight);
        Point begin = new Point();
        begin.set(ALIGN_PADDING_TOP,y);
        Point end = new Point();
        end.set(getWidth() - ALIGN_PADDING_RIGHT,y);
        ILog.e(TAG,"begin:"+begin+" end:"+end);
        drawDottedLine(canvas, begin, end, mTargetPaint);
        canvas.drawText(mContext.getResources().getString(R.string.target_weight), end.x - 120, end.y + 40, mTargetTextPaint);
/*        ILog.e(TAG, "drawTargetLine");
        Data begin = new Data();
        begin.setWeight(mDataAdapter.getTargetWeight());
        begin.setTime(mDataAdapter.getTimeSmallest());

        Data end = new Data();
        end.setWeight(mDataAdapter.getTargetWeight());
        end.setTime(mDataAdapter.getTimeBiggest());

        int beginId = mDataAdapter.getShowDataStartId();
        float steps = mWidth / (mDataAdapter.getShowPointCount()-1);
        Point beginPoint = toViewPoint(begin, 0, steps);
        Point endPoint = toViewPoint(end,mDataAdapter.getShowPointCount()-1 , steps);
        drawDottedLine(canvas, beginPoint, endPoint, mTargetPaint);
        ILog.d(TAG, "yanlog drawTargetLine beginPoint:" + beginPoint + " endPoint:" + endPoint);
        ILog.d(TAG, "yanlog targetLine begin:" + begin + " :end:" + end);

        canvas.drawText(mContext.getResources().getString(R.string.target_weight), endPoint.x - 120, endPoint.y + 40, mTargetTextPaint);*/
        /*canvas.drawLine(beginPoint.x, beginPoint.y, endPoint.x, endPoint.y, mTargetPaint);*/
    }

    /**
     * 画各个体重节点
     *
     * @param canvas
     */
    private void drawPoint(Canvas canvas) {
        if (mDataAdapter == null || mDataAdapter.getShowPointCount() == 0) {
            return;
        }
        if (mDataAdapter.getShowDataList().size() == 1) {
/*            canvas.drawCircle(ALIGN_PADDING_LEFT,
                    ALIGN_PADDING_TOP + mHeight / 2 10,mPointPaint);*/
            canvas.drawCircle(ALIGN_PADDING_LEFT,
                    toYPoint(mDataAdapter.getShowDataList().get(0).getWeight()), 10,mPointPaint);
            return;
        }
        //int beginId = mDataAdapter.getShowDataStartId();
        float steps = mWidth / (mDataAdapter.getShowPointCount()-1);
        ArrayList<Data> list = mDataAdapter.getShowDataList();
        int i = 0;
        for (Data data : list) {
            Point point = toViewPoint(data, i++, steps);
            canvas.drawCircle(point.x, point.y, 10, mPointPaint);
        }
    }

    private float mLastTouchPos = 0;
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float curTouchPos = event.getX();
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:

                break;
            case MotionEvent.ACTION_UP:

                break;
            case MotionEvent.ACTION_MOVE:
                mCurMoveLength += curTouchPos - mLastTouchPos;
                mDataAdapter.notifyDataPosSetChange(mCurMoveLength/mWidth * 7);
                break;
            default:
                break;
        }
        mLastTouchPos = curTouchPos;
        return super.onTouchEvent(event);
    }

    /**
     * 画图下面的时间
     *
     * @param canvas
     */
    private void drawTimeText(Canvas canvas) {
        if (mDataAdapter == null) {
            return;
        }
        long minTime = mDataAdapter.getTimeSmallest();
        long maxTime = mDataAdapter.getTimeBiggest();
        canvas.drawText(OldUtils.formatTime(minTime), ALIGN_PADDING_LEFT, 40 + ALIGN_PADDING_TOP + mLineSpacing * SPACES_COUNT,
                mTextPaint);
        canvas.drawText(OldUtils.formatTime((minTime + maxTime) / 2), ALIGN_PADDING_LEFT + mWidth / 2 - 60, 40 + ALIGN_PADDING_TOP + mLineSpacing * SPACES_COUNT,
                mTextPaint);
        canvas.drawText(OldUtils.formatTime(maxTime), ALIGN_PADDING_LEFT + mWidth - 110, 40 + ALIGN_PADDING_TOP + mLineSpacing * SPACES_COUNT,
                mTextPaint);

    }

    /**
     * 画y轴上的体重数字
     *
     * @param canvas
     */
    private void drawWeightText(Canvas canvas) {
        if (mDataAdapter == null) {
            return;
        }
        float maxvalue = mDataAdapter.getWeightBiggest();
        float space = mDataAdapter.getHeight() / SPACES_COUNT;
        canvas.drawText(String.format("%.1f", maxvalue), ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP + 40, mTextPaint);
        for (int i = 1; i < SPACES_COUNT; i++) {
            canvas.drawText(String.format("%.1f", maxvalue - i * space),
                    ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP + 40 + i * mLineSpacing, mTextPaint);
        }
    }


    /**
     * 画中间的虚线
     *
     * @param canvas
     */
    private void drawDottedLine(Canvas canvas) {
        Path path = new Path();
        int space = 10;
        path.moveTo(mWidth / 2 + ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP);
        for (int i = 1; i * space < mHeight; i += 2) {
            path.lineTo(mWidth / 2 + ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP + i * space);
            path.moveTo(mWidth / 2 + ALIGN_PADDING_LEFT, ALIGN_PADDING_TOP + (i + 1) * space);
        }
        canvas.drawPath(path, mDottedLinePaint);
    }

    /**
     * 画虚线的一个共性的方法
     *
     * @param canvas
     * @param begin
     * @param end
     * @param paint
     */
    private void drawDottedLine(Canvas canvas, Point begin, Point end, Paint paint) {
        float height = end.y - begin.y;
        float width = end.x - begin.x;
        if (Math.abs(width) > Math.abs(height)) {
            Path path = new Path();
            int space = 10;
            float spaceX = space;

            if (end.x < begin.x) {
                spaceX = -space;
            }
            float spaceY = ((float) spaceX) / width * height;
            path.moveTo(begin.x, begin.y);
            for (int i = 1; i * space < Math.abs(width); i += 2) {
                path.lineTo(begin.x + i * spaceX, begin.y + i * spaceY);
                path.moveTo(begin.x + (i + 1) * spaceX, begin.y + (i + 1) * spaceY);
            }
            canvas.drawPath(path, paint);
        } else {
            Path path = new Path();
            int space = 10;
            float spaceY = space;

            if (end.y < begin.y) {
                spaceY = -space;
            }
            float spaceX = ((float) spaceY) / height * width;
            path.moveTo(begin.x, begin.y);
            for (int i = 1; i * space < Math.abs(height); i += 2) {
                path.lineTo(begin.x + i * spaceX, begin.y + i * spaceY);
                path.moveTo(begin.x + (i + 1) * spaceX, begin.y + (i + 1) * spaceY);
            }
            canvas.drawPath(path, paint);
        }

    }

    /**
     * 画体重的折线
     *
     * @param canvas
     */
    private void drawPolyLine(Canvas canvas) {
        ArrayList<Data> list = mDataAdapter.getShowDataList();
        if (list == null || list.size() == 0 || list.size() == 1) {
            return;
        }
        Path path = new Path();
        int beginsId = mDataAdapter.getShowDataStartId();
        float steps = mWidth / (mDataAdapter.getShowPointCount()-1);
        Point begin = toViewPoint(list.get(0), 0, steps);
        ILog.e(TAG,"drawPolyLine:"+begin.toString()+":"+list.get(0).toString());
        Point end = null;
        path.moveTo(begin.x, begin.y);
        Point fixPoint_1 = new Point();
        Point fixPoint_2 = new Point();
        for (int i = 1; i < list.size(); i++) {
            end = toViewPoint(list.get(i), i, steps);
            ILog.e(TAG,"drawPolyLine:"+end.toString()+":"+list.get(i).toString());
            fixPoint_1.set((begin.x + end.x) / 2, begin.y);
            fixPoint_2.set((begin.x + end.x) / 2, end.y);
            path.cubicTo(fixPoint_1.x, fixPoint_1.y, fixPoint_2.x, fixPoint_2.y, end.x, end.y);
            begin = end;
        }
        canvas.drawPath(path, mPolyLinePaint);
    }

    /**
     * 将各个体重值转换为在view中的坐标值
     *
     * @param data
     * @return
     */
    private Point toViewPoint(Data data, int ids, float timeSteps) {
        if (mDataAdapter == null) {
            return null;
        }
        Point point = new Point();
        point.set(ALIGN_PADDING_LEFT + (int) (ids * timeSteps), ALIGN_PADDING_TOP + (int) (((mDataAdapter.getWeightBiggest() - data.getWeight())
                * (float) mHeight / (mDataAdapter.getHeight()))));
        return point;
    }

    private int toYPoint(float weight){
        return  ALIGN_PADDING_TOP + (int) (((mDataAdapter.getWeightBiggest() - weight)
                * (float) mHeight / (mDataAdapter.getHeight())));
    }

}

File: app/src/main/java/com/plbear/iweight/model/main/view/LineChartView.kt
package com.plbear.iweight.model.main.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Point
import android.graphics.Typeface
import android.os.Handler
import android.os.Message
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

import com.plbear.iweight.data.DataManager
import com.plbear.iweight.R
import com.plbear.iweight.data.Data
import com.plbear.iweight.model.main.adapter.LineChartAdapter
import com.plbear.iweight.utils.MyLog
import com.plbear.iweight.utils.Utils

/**
 * Created by koakira on 16/11/normal_5.
 */

class LineChartView(context: Context, attributeSet: AttributeSet) : View(context, attributeSet) {
    private var TAG = "LineChartView:"
    private val SPACES_COUNT = 5
    private val MSG_DATA_CHANGE = 0
    private val ALIGN_PADDING_LEFT = 20
    private val ALIGN_PADDING_RIGHT = 20
    private val ALIGN_PADDING_TOP = 10
    private val ALIGN_PADDING_BOTTOM = 40

    private  var mBondLinePaint: Paint = Paint()
    private  var mLinePaint: Paint = Paint()
    private  var mDottedLinePaint: Paint = Paint()
    private  var mPolyLinePaint: Paint = Paint()
    private  var mTextPaint: Paint = Paint()
    private  var mPointPaint: Paint = Paint()
    private  var mTargetPaint: Paint = Paint()
    private  var mTargetTextPaint: Paint = Paint()
    private  var mDataManager: DataManager = DataManager.getInstance()

    private var mTop = 0
    private var mLeft = 0
    private var mRight = 0
    private var mBottom = 0
    private var mWidth = 0
    private var mHeight = 0
    private var mLineSpacing = 0
    private var mCurMoveLength = 0


    /*public void setDataAdapter(LineChartAdapter adapter) {
        mDataAdapter = adapter;
        Message msg = mHandler.obtainMessage(MSG_DATA_CHANGE);
        mHandler.sendMessage(msg);
    }*/
    var dataAdpater = LineChartAdapter()
        private set
    private var mContext: Context? = null
    private val mHandler = object : Handler() {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                MSG_DATA_CHANGE -> {
                    MyLog.d(TAG, "Msg_data_change")
                    this@LineChartView.invalidate()
                }
                else -> {
                }
            }
        }
    }

    private var mLastTouchPos = 0f

    fun setTag(tag: String) {
        TAG += tag
        dataAdpater!!.setTag(TAG)
    }

    init {
        mContext = context
        init()
    }

    /**
     * 初始化画笔工具
     */
    private fun initPaints() {
        mBondLinePaint.color = Color.BLACK
        mBondLinePaint.typeface = Typeface.DEFAULT_BOLD
        mBondLinePaint.style = Paint.Style.STROKE
        mBondLinePaint.strokeWidth = 3f
        mBondLinePaint.isAntiAlias = true
        mBondLinePaint.alpha = 130

        mLinePaint.color = Color.BLACK
        mLinePaint.style = Paint.Style.STROKE
        mLinePaint.strokeWidth = 3f
        mLinePaint.alpha = 90
        mLinePaint.isAntiAlias = true

        mDottedLinePaint.color = Color.BLACK
        mDottedLinePaint.style = Paint.Style.STROKE
        mDottedLinePaint.strokeWidth = 3f
        mDottedLinePaint.isAntiAlias = true
        mDottedLinePaint.alpha = 150

        mPolyLinePaint.color = resources.getColor(R.color.line_color)
        mPolyLinePaint.style = Paint.Style.STROKE
        mPolyLinePaint.strokeWidth = 5f
        mPolyLinePaint.isAntiAlias = true

        mTextPaint.color = Color.BLACK
        mTextPaint.textSize = 38f
        mTextPaint.isAntiAlias = true
        mTextPaint.style = Paint.Style.FILL
        mTextPaint.alpha = 150
        mTextPaint.strokeWidth = 1f

        mPointPaint.color = resources.getColor(R.color.point_color)
        mPointPaint.strokeWidth = 20f

        mTargetPaint.color = resources.getColor(R.color.target_weight)
        mTargetPaint.style = Paint.Style.STROKE
        mTargetPaint.strokeWidth = 4f
        mTargetPaint.isAntiAlias = true

        mTargetTextPaint.color = resources.getColor(R.color.target_weight)
        mTargetTextPaint.textSize = 38f
        mTargetTextPaint.isAntiAlias = true
        mTargetTextPaint.style = Paint.Style.FILL
        mTargetTextPaint.alpha = 150
        mTargetTextPaint.strokeWidth = 1f

    }

    private fun init() {
        MyLog.e(TAG, "init")
        dataAdpater!!.setTag(TAG)
        dataAdpater!!.registerDataListener(object : LineChartAdapter.DataChangeListener {
            override fun onChange() {
                MyLog.d(TAG, "LineChartView data changed")
                val msg = mHandler.obtainMessage(MSG_DATA_CHANGE)
                mHandler.sendMessage(msg)
            }
        })
        initPaints()
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        if (changed) {
            mLeft = left
            mTop = top
            mRight = right
            mBottom = bottom
            mWidth = mRight - mLeft - ALIGN_PADDING_LEFT - ALIGN_PADDING_RIGHT
            mHeight = mBottom - mTop - ALIGN_PADDING_TOP - ALIGN_PADDING_BOTTOM
            MyLog.d(TAG, "mLeft:$mLeft mRight:$mRight mTop:$mTop mBottom:$mBottom mWeight:$mWidth mHeight:$mHeight")
            mLineSpacing = (mBottom - mTop - ALIGN_PADDING_BOTTOM - ALIGN_PADDING_TOP) / SPACES_COUNT
        }
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawLine(ALIGN_PADDING_LEFT.toFloat(), ALIGN_PADDING_TOP.toFloat(), (mRight - ALIGN_PADDING_RIGHT).toFloat(),
                ALIGN_PADDING_TOP.toFloat(), mBondLinePaint)
        for (i in 1 until SPACES_COUNT) {
            canvas.drawLine(ALIGN_PADDING_LEFT.toFloat(), (ALIGN_PADDING_TOP + i * mLineSpacing).toFloat(), (mRight - ALIGN_PADDING_RIGHT).toFloat(), (ALIGN_PADDING_TOP + i * mLineSpacing).toFloat(), mLinePaint!!)
        }
        canvas.drawLine(ALIGN_PADDING_LEFT.toFloat(), (ALIGN_PADDING_TOP + SPACES_COUNT * mLineSpacing).toFloat(), (mRight - ALIGN_PADDING_RIGHT).toFloat(), (ALIGN_PADDING_TOP + SPACES_COUNT * mLineSpacing).toFloat(), mBondLinePaint!!)
        drawDottedLine(canvas)
        drawPolyLine(canvas)
        drawWeightText(canvas)
        drawTimeText(canvas)
        val temp = dataAdpater!!.showPointCount
        if (dataAdpater!!.showPointCount < 50) {
            drawPoint(canvas)
        }
        drawTargetLine(canvas)

        super.onDraw(canvas)
    }


    /**
     * 画目标体重的线和“目标值”这几个数值
     *
     * @param canvas
     */
    private fun drawTargetLine(canvas: Canvas) {
        val targetWeight = dataAdpater!!.targetWeight
        if (targetWeight == -1f) {
            MyLog.e(TAG, "target weight is not set return")
            return
        }
        val y = toYPoint(targetWeight)
        val begin = Point()
        begin.set(ALIGN_PADDING_TOP, y)
        val end = Point()
        end.set(width - ALIGN_PADDING_RIGHT, y)
        MyLog.e(TAG, "begin:$begin end:$end")
        drawDottedLine(canvas, begin, end, mTargetPaint)
        canvas.drawText(mContext!!.resources.getString(R.string.target_weight), (end.x - 120).toFloat(), (end.y + 40).toFloat(), mTargetTextPaint!!)
    }

    /**
     * 画各个体重节点
     *
     * @param canvas
     */
    private fun drawPoint(canvas: Canvas) {
        if (dataAdpater == null || dataAdpater!!.showPointCount == 0) {
            return
        }
        if (dataAdpater!!.showDataList.size == 1) {
            /*            canvas.drawCircle(ALIGN_PADDING_LEFT,
                    ALIGN_PADDING_TOP + mHeight / normal_2 10,mPointPaint);*/
            canvas.drawCircle(ALIGN_PADDING_LEFT.toFloat(),
                    toYPoint(dataAdpater!!.showDataList[0].weight).toFloat(), 10f, mPointPaint!!)
            return
        }
        //int beginId = mDataAdapter.getShowDataStartId();
        val steps = (mWidth / (dataAdpater!!.showPointCount - 1)).toFloat()
        val list = dataAdpater!!.showDataList
        var i = 0
        for (data in list) {
            val point = toViewPoint(data, i++, steps)
            canvas.drawCircle(point!!.x.toFloat(), point.y.toFloat(), 10f, mPointPaint!!)
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val curTouchPos = event.x
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
            }
            MotionEvent.ACTION_UP -> {
            }
            MotionEvent.ACTION_MOVE -> {
                mCurMoveLength += (curTouchPos - mLastTouchPos).toInt()
                dataAdpater!!.notifyDataPosSetChange((mCurMoveLength / mWidth * 7).toFloat())
            }
            else -> {
            }
        }
        mLastTouchPos = curTouchPos
        return super.onTouchEvent(event)
    }

    /**
     * 画图下面的时间
     *
     * @param canvas
     */
    private fun drawTimeText(canvas: Canvas) {
        val minTime = dataAdpater.timeSmallest
        val maxTime = dataAdpater.timeBiggest
        canvas.drawText(Utils.formatTime(minTime), ALIGN_PADDING_LEFT.toFloat(), (40 + ALIGN_PADDING_TOP + mLineSpacing * SPACES_COUNT).toFloat(),
                mTextPaint)
        canvas.drawText(Utils.formatTime((minTime + maxTime) / 2), (ALIGN_PADDING_LEFT + mWidth / 2 - 60).toFloat(), (40 + ALIGN_PADDING_TOP + mLineSpacing * SPACES_COUNT).toFloat(),
                mTextPaint)
        canvas.drawText(Utils.formatTime(maxTime), (ALIGN_PADDING_LEFT + mWidth - 110).toFloat(), (40 + ALIGN_PADDING_TOP + mLineSpacing * SPACES_COUNT).toFloat(),
                mTextPaint)

    }

    /**
     * 画y轴上的体重数字
     *
     * @param canvas
     */
    private fun drawWeightText(canvas: Canvas) {
        if (dataAdpater == null) {
            return
        }
        val maxvalue = dataAdpater!!.weightBiggest
        val space = dataAdpater!!.height / SPACES_COUNT
        canvas.drawText(String.format("%.1f", maxvalue), ALIGN_PADDING_LEFT.toFloat(), (ALIGN_PADDING_TOP + 40).toFloat(), mTextPaint)
        for (i in 1 until SPACES_COUNT) {
            canvas.drawText(String.format("%.1f", maxvalue - i * space),
                    ALIGN_PADDING_LEFT.toFloat(), (ALIGN_PADDING_TOP + 40 + i * mLineSpacing).toFloat(), mTextPaint)
        }
    }


    /**
     * 画中间的虚线
     *
     * @param canvas
     */
    private fun drawDottedLine(canvas: Canvas) {
        val path = Path()
        val space = 10
        path.moveTo((mWidth / 2 + ALIGN_PADDING_LEFT).toFloat(), ALIGN_PADDING_TOP.toFloat())
        var i = 1
        while (i * space < mHeight) {
            path.lineTo((mWidth / 2 + ALIGN_PADDING_LEFT).toFloat(), (ALIGN_PADDING_TOP + i * space).toFloat())
            path.moveTo((mWidth / 2 + ALIGN_PADDING_LEFT).toFloat(), (ALIGN_PADDING_TOP + (i + 1) * space).toFloat())
            i += 2
        }
        canvas.drawPath(path, mDottedLinePaint!!)
    }

    /**
     * 画虚线的一个共性的方法
     *
     * @param canvas
     * @param begin
     * @param end
     * @param paint
     */
    private fun drawDottedLine(canvas: Canvas, begin: Point, end: Point, paint: Paint?) {
        val height = (end.y - begin.y).toFloat()
        val width = (end.x - begin.x).toFloat()
        if (Math.abs(width) > Math.abs(height)) {
            val path = Path()
            val space = 10
            var spaceX = space.toFloat()

            if (end.x < begin.x) {
                spaceX = (-space).toFloat()
            }
            val spaceY = spaceX.toFloat() / width * height
            path.moveTo(begin.x.toFloat(), begin.y.toFloat())
            var i = 1
            while (i * space < Math.abs(width)) {
                path.lineTo(begin.x + i * spaceX, begin.y + i * spaceY)
                path.moveTo(begin.x + (i + 1) * spaceX, begin.y + (i + 1) * spaceY)
                i += 2
            }
            canvas.drawPath(path, paint)
        } else {
            val path = Path()
            val space = 10
            var spaceY = space.toFloat()

            if (end.y < begin.y) {
                spaceY = (-space).toFloat()
            }
            val spaceX = spaceY.toFloat() / height * width
            path.moveTo(begin.x.toFloat(), begin.y.toFloat())
            var i = 1
            while (i * space < Math.abs(height)) {
                path.lineTo(begin.x + i * spaceX, begin.y + i * spaceY)
                path.moveTo(begin.x + (i + 1) * spaceX, begin.y + (i + 1) * spaceY)
                i += 2
            }
            canvas.drawPath(path, paint!!)
        }

    }

    /**
     * 画体重的折线
     *
     * @param canvas
     */
    private fun drawPolyLine(canvas: Canvas) {
        val list = dataAdpater.showDataList
        if (list.size == 0 || list.size == 1) {
            return
        }
        val path = Path()
        val beginsId = dataAdpater.showDataStartId
        val steps = (mWidth / (dataAdpater.showPointCount - 1)).toFloat()
        var begin = toViewPoint(list[0], 0, steps)
        var end: Point? = null
        path.moveTo(begin.x.toFloat(), begin.y.toFloat())
        val fixPoint_1 = Point()
        val fixPoint_2 = Point()
        for (i in 1 until list.size) {
            end = toViewPoint(list[i], i, steps)
            fixPoint_1.set((begin.x + end.x) / 2, begin.y)
            fixPoint_2.set((begin.x + end.x) / 2, end.y)
            path.cubicTo(fixPoint_1.x.toFloat(), fixPoint_1.y.toFloat(), fixPoint_2.x.toFloat(), fixPoint_2.y.toFloat(), end.x.toFloat(), end.y.toFloat())
            begin = end
        }
        canvas.drawPath(path, mPolyLinePaint!!)
    }

    /**
     * 将各个体重值转换为在view中的坐标值
     *
     * @param data
     * @return
     */
    private fun toViewPoint(data: Data, ids: Int, timeSteps: Float): Point {
        val point = Point()
        point.set(ALIGN_PADDING_LEFT + (ids * timeSteps).toInt(), ALIGN_PADDING_TOP + ((dataAdpater.weightBiggest - data.weight) * mHeight.toFloat() / dataAdpater.height).toInt())
        return point
    }

    private fun toYPoint(weight: Float): Int {
        return ALIGN_PADDING_TOP + ((dataAdpater.weightBiggest - weight) * mHeight.toFloat() / dataAdpater.height).toInt()
    }

}
